compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java

```diff
@@ -549,6 +571,54 @@ private boolean isValidReturnBType(Class<?> jType, BType bType, JMethodRequest j
                 this.symbolTable.arrayJsonType};
     }
 
+    private boolean isReadOnlyCompatibleReturnType(Class<?> jType, JMethodRequest jMethodRequest)
+            throws ClassNotFoundException {
+        if (jType.getTypeName().equals(J_OBJECT_TNAME)) {
+            return true;
+        }
+
+        for (BType member : definedReadOnlyMemberTypes) {
+            if (isValidReturnBType(jType, member, jMethodRequest)) {
+                return true;
+            }
+        }
+
+        if (this.classLoader.loadClass(BError.class.getCanonicalName()).isAssignableFrom(jType)) {
+            return true;
+        }
+
+        if (this.classLoader.loadClass(BFunctionPointer.class.getCanonicalName()).isAssignableFrom(jType)) {
+            return true;
+        }
+
+        if (this.classLoader.loadClass(BObject.class.getCanonicalName()).isAssignableFrom(jType)) {
+            // Service and Object.
+            return true;
+        }
+
+        if (this.classLoader.loadClass(BTypedesc.class.getCanonicalName()).isAssignableFrom(jType)) {
+            return true;
+        }
+
+        if (this.classLoader.loadClass(BHandle.class.getCanonicalName()).isAssignableFrom(jType)) {
+            return true;
+        }
+
+        if (this.classLoader.loadClass(BXML.class.getCanonicalName()).isAssignableFrom(jType)) {
+            return true;
+        }
+
+        if (this.isValidListType(jType, jMethodRequest)) {
+            return true;
+        }
+
+        if (this.classLoader.loadClass(BMap.class.getCanonicalName()).isAssignableFrom(jType)) {
+            return true;
+        }
```

Comment:
Can't we combine the `return true` if statements using `||` ? and move all checks or individual checks to a different method if it makes sense

Reply:
Changed in 28cb8ed.